[wip]: FontDownloader plugin

pazos opened the pull request:
Tries to showcase my proposal in #6272 (and the feedback received that some gfonts were not optimal).

It is still very much a draft, but shows how to display a few fonts as featured and how to use our own alternatives instead of google fonts. At the same time makes available the Google Foo for new users.

Fonts can be navigated by: featured, language (glyph support), category (sans, serif, display...), date of last modification (last week, month or year, older) and the whole ~1000 fonts alphabetically.

Since the font table is tiny(1MB) we store that in the cache and just require a network connection when we're going to download fonts. On download all variants for a given font are downloaded.

And yeah, the look & feel is basically the same that calibre search metadata, because it is a low-effort copy!!!

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/6299)
<!-- Reviewable:end -->

NiLuJe commented on plugins/gfonts.koplugin/featured.lua:
```diff
@@ -0,0 +1,61 @@
+
+    -- we'll use custom versions of the following font families:
+    own_fonts = {
+	-- NiLuJe build of Literata 3 goes here
```
I don't think the license allows redistribution of modified files?

NiLuJe replied:
Well, not exactly. It's OFL, so, it can, but not under the name "Literata". Incidentally, I'm currently renaming it to "Lit3rata" (originally purely to be able to compare it to Literata 2 :D). Dunno if that's too cheap to count ;p.

pazos replied:
I see no problem with SIL OFL. See https://theleagueof.github.io/licenses/ofl-faq.html (specially chapter 3)

NiLuJe replied:
In which case:

https://storage.sbg.cloud.ovh.net/v1/AUTH_2ac4bfee353948ec8ea7fd1710574097/koreader-pub/Lit3rata-Regular.otf
https://storage.sbg.cloud.ovh.net/v1/AUTH_2ac4bfee353948ec8ea7fd1710574097/koreader-pub/Lit3rata-Italic.otf
https://storage.sbg.cloud.ovh.net/v1/AUTH_2ac4bfee353948ec8ea7fd1710574097/koreader-pub/Lit3rata-Bold.otf
https://storage.sbg.cloud.ovh.net/v1/AUTH_2ac4bfee353948ec8ea7fd1710574097/koreader-pub/Lit3rata-BoldItalic.otf

(I'd add a link to http://trac.ak-team.com/trac/browser/niluje/Configs/trunk/Kindle/Kobo_Hacks/Patches/REMINDER#L125 in the comments).

And, FWIW, I'm using Regular for Regular, not Medium ;).

NiLuJe replied:
The metadata looks sane for a Modified Version, if I compare it to the Bitter/BitterPro situation ;).

pazos replied:
Included in last commit. It does look nice!!!

NiLuJe commented on plugins/gfonts.koplugin/featured.lua:
```diff
@@ -0,0 +1,61 @@
+local Featured = {
+    -- recommended fonts, both from gfonts and custom repos
+    fonts = {
+	"Bitter",
```
Indent ;).

Frenzie commented on plugins/gfonts.koplugin/api.lua:
```diff
@@ -0,0 +1,89 @@
+
+function Api:request()
+    local request, sink = {}, {}
+    request['url'] = self.base_url .. "?key=" .. self.api_key
```
No real reason to use request["url"] in places like this. (Unless you really want to I suppose.)
```suggestion
    request.url = self.base_url .. "?key=" .. self.api_key
```

Frenzie commented on plugins/gfonts.koplugin/api.lua:
```diff
@@ -0,0 +1,89 @@
+function Api:request()
+    local request, sink = {}, {}
+    request['url'] = self.base_url .. "?key=" .. self.api_key
+    request['method'] = 'GET'
```
```suggestion
    request.method = "GET"
```

Frenzie commented on plugins/gfonts.koplugin/featured.lua:
```diff
@@ -0,0 +1,61 @@
+local Featured = {
+    -- recommended fonts, both from gfonts and custom repos
```
Recommended by you? :-)

pazos replied:
I took samples from https://www.mobileread.com/forums/showthread.php?t=316565 and the referenced issue. Some thread recommendations are not available on gfonts (tex gyrella, linux libertine...)

Frenzie commented on plugins/gfonts.koplugin/main.lua:
```diff
@@ -0,0 +1,283 @@
+function GFonts:downloadFont(family)
+    logger.info("Prompt download " .. family)
+    UIManager:show(ConfirmBox:new{
+        text = T(_("Do you want to download %1"), family),
```
Maybe short and to the point? Either way I'd probably add that question mark.
```suggestion
        text = T(_("Download %1?"), family),
```

yparitcher commented on plugins/gfonts.koplugin/api.lua:
```diff
@@ -0,0 +1,89 @@
+function Api:init(file)
+    local fallback_key = "AIzaSyDQZaihK8Lb7jJ3DYyrQrpyyJF7tyvrqAs"
+    if not file then
+        self.api_key = fallback_key
```
A simpler way to rewrite this function
```
    if file then
        local keyfile = io.open(file, "r")
        if keyfile then
            local key = keyfile:read("*a")
            keyfile:close()
            self.api_key = key
            return
        end
    end
    -- fallback key
    self.api_key = "AIzaSyDQZaihK8Lb7jJ3DYyrQrpyyJF7tyvrqAs"
```

pazos commented:
I have an issue with the Confirmbox I'm not sure how to fix. Its ok_callback has the code that downloads all font variants for a given family, so it blocks until all files are downloaded and all infomessages are skipped except the last one.

I've tried to send events instead of calling the function to download files but the result is the same.

poire-z commented:
> so it blocks until all files are downloaded and all infomessages are skipped except the last one.

What are you trying to achieve? Some kind of progress, each InfoMessage showing `Downloading font xxx 1/4` ?
If yes, you'd need to force a repaint after showing it, as UIManager won't find a time slot to do its usual aggregative painting and refreshes.
The other option is to use `Trapper:info()`, which is supposed to handle all that behind the scene and make it easier for that kind of UI progress stuff (but it's a bit tedious to understand :/ using coroutines internally...)
The Wikipedia `Wikipedia:createEpubWithUI()` and CoverBrowser's `BookInfoManager:extractBooksInDirectory()` use it. It looks also ideal for you if I understand your need right (and it allows interrupting between downloads).

pazos commented:
> What are you trying to achieve? Some kind of progress, each InfoMessage showing Downloading font xxx 1/4 ?
If yes, you'd need to force a repaint after showing it, as UIManager won't find a time slot to do its usual aggregative painting and refreshes.

Yeah, using `UIManager:forceRePaint()` works fine. Thank you!

pazos commented:
Well, with lastest changes in this PR is ready to be tested on the emulator. 

Just two things to do, but they're easy:
- actually download fonts in the font_dir (just logged, but is still downloaded in current dir)
- network availability checks

Now just the difficult part of it:

- name of the plugin?.

It is mainly a gfonts downloader (981 of the 982 fonts are from google), but that can change in the future.

- where to place the menus?

poire-z commented:
`fontdownloader.plugin/` , `Font downloader` ( or `Fontastic downloader` :)
I guess it could add itself in `Font settings>` as `Download fonts`, alongside `Enable system fonts` and `Open fonts folder` (I guess we won't see it when we disable the plugin?)

Frenzie commented:
"Pile of typography" :-P

(Font downloader is the most obvious choice.)

pazos commented:
Here are some screenshots. Any suggestion instead of "languages"?. It should be saying "unicode coverage subsets" but that sounds too technical.

![2020-06-25_17-06](https://user-images.githubusercontent.com/975883/85746303-a5594800-b706-11ea-9da8-a3bc7f78e325.png)

![2020-06-25_17-06_1](https://user-images.githubusercontent.com/975883/85746346-ae4a1980-b706-11ea-9edb-820eb36f7a94.png)

![2020-06-25_17-05](https://user-images.githubusercontent.com/975883/85746420-bc983580-b706-11ea-8ea6-3fd1f8a3a7b4.png)

Frenzie commented:
> Any suggestion instead of "languages"?

Would "scripts" be accurate?

poire-z commented:
> Would "scripts" be accurate?

I think it is, but it's a bit less obvious than "languages" for most people I guess.
May be `Scripts (languages)` would be ok?

poire-z commented:
And may be `Download more fonts` or `Download additional fonts` for the menu.

pazos commented:
Addressed most of @poire-z suggestions. Plugin renamed to `fontdownloader.koplugin` (albeit I do really like `fontastic.koplugin` .The "Download more fonts" menu entry is injected in the font settings submenu for CRE documents but I see no problem in using a shortcut to invoke the action from other documents or even the FM.

Still not sure about the languages/scripts dilema. I think we can leave languages in the front search menu (most obvious) and change to "script support" or "unicode support" in the details (3rd screenshot).

BTW, is there any way to debug wireless/network dependent code from the emulator?. Can we simulate a "network is down" state?. It is useful to test if network checks and actions are triggered correctly.

poire-z commented on plugins/fontdownloader.koplugin/main.lua:
```diff
@@ -0,0 +1,514 @@
+    fonts = {},
+    blacklist = {
+        "Noto Sans", "Noto Sans HK", "Noto Sans JP", "Noto Sans KR", "Noto Sans SC", "Noto Sans TC",
+        "Noto Serif", "Noto Serif HK", "Noto Serif JP", "Noto Serif KR", "Noto Serif SC", "Noto Serif TC",
```
We don't ship any `Noto Serif CJK` - so may be you could unblacklist `Noto Serif SC` - assuming:
- CJK users really care about Sans vs Serif :)
- Serif SC includes the TC/KR/JP variations like our NotoSansCJKsc does, which will be correctly picked up by the recent typography stuff

pazos replied:
> Serif SC includes the TC/KR/JP variations like our NotoSansCJKsc does, which will be correctly picked up by the recent typography stuff

Never checked but it looks like is splited in "Noto Serif HK", "Noto Serif JP", "Noto Serif KR", "Noto Serif SC" and "Noto Serif TC". So we need to unblacklist all of them.

poire-z replied:
https://www.google.com/get/noto/help/cjk/ states that:
> Each font sets one language as the default. Note that each language-specific font **does** support all four languages and includes the complete set of glyphs

I don't know which thing your plugin will download, but you can try to see if the 5 are proposed in the list of font supporting the Korean script?
And/or download one of them (SC, for consistency), and use the `Generate font test document` and go to that font, and try to toggle `Typography rules>` between chinese and japanese: the few CJK chars in that sample should slightly change (check they do with Noto Sans CJK).

pazos replied:
Nope. Each noto serif fonts support two subsets: latin and its own (based on names).
![2020-06-26_16-53](https://user-images.githubusercontent.com/975883/85870646-98ebf280-b7cd-11ea-81a9-ce1033bb5d97.png)

It makes some sense as the api is intended mostly to provide font faces to webpages and there's no need to send the whole CJK when the page is written in just one language.

Other thing to consider is to release our own Noto Serif CJK somewhere and include it like the Lit3rata font.

Frenzie commented:
At least in Dutch I don't think *schrift* (script) is particularly niche. In English it might be better to translate it as writing system.

poire-z commented:
I see the menu twice on the emulator:
<kbd>![image](https://user-images.githubusercontent.com/24273478/85872924-b8384f00-b7d0-11ea-8219-98997e58c79a.png)</kbd>

poire-z commented on plugins/fontdownloader.koplugin/main.lua:
```diff
@@ -0,0 +1,514 @@
+    name = "fontdownloader",
+    is_doc_only = false,
+    base_url = "https://www.googleapis.com/webfonts/v1/webfonts",
+    user_cache = DataStorage:getDataDir() .. "/cache/gbooks.lua",
+    user_key = DataStorage:getSettingsDir() .. "gfonts-api.txt",
```
Might be changed too to the plugin final name ? At lease the cache.

poire-z commented:
On the first hit on that menu, I see a PING in the logs and nothing else - and the gbooks.lua is created. Only on next menu hit I see the popup. Is that ok ?

poire-z commented:
Some crash, that I didn't get on my previous playing, but after I had removed gbooks.lua to get it regenerated without Noto Serif blacklisted, when I hit "Family":
```
./luajit: plugins/fontdownloader.koplugin/main.lua:367: attempt to index local 'id' (a nil value)
stack traceback:
        plugins/fontdownloader.koplugin/main.lua:367: in function 'info'
        plugins/fontdownloader.koplugin/main.lua:375: in function 'fontCatalog'
        plugins/fontdownloader.koplugin/main.lua:335: in function 'find'
        plugins/fontdownloader.koplugin/main.lua:308: in function 'close'
        plugins/fontdownloader.koplugin/main.lua:263: in function 'callback'
        frontend/ui/widget/button.lua:222: in function 'handleEvent'
```

poire-z commented:
It has downloaded:
```
Noto Serif SC-regular.ttf
Noto Serif SC-bold.ttf
Noto Serif SC-medium.ttf
```

(OK, I have to manually moved them to the font dir.)
But shouldn't it download only regular OR medium ? Downloading both will confuse crengine, which will pick the first one, probably medium ?

pazos commented:
> Some crash, that I didn't get on my previous playing, but after I had removed gbooks.lua to get it regenerated without Noto Serif blacklisted, when I hit "Family":

Fixed

> (OK, I have to manually moved them to the font dir.)

Yup, thats not implented :p . I'm going to do it now because I see no reason to wait.

> But shouldn't it download only regular OR medium ? Downloading both will confuse crengine, which will pick the first one, probably medium ?

Font variants that we want are declared in the `variants` table:

```lua
    variants = {
        ["regular"]            = "regular",
        ["italic"]             = "italic",
        ["500"]                = "medium",
        ["500italic"]          = "medium-italic",
        ["700"]                = "bold",
        ["700italic"]          = "bold-italic",
    },
```

We can change them as we like, so probably makes sense to avoid medium completely.

poire-z commented:
OK, confirming Noto Serif SC only has SC glyphs:
<kbd>![image](https://user-images.githubusercontent.com/24273478/85874730-61804480-b7d3-11ea-9709-83572ee3049e.png)</kbd>
while Noto Sans CJK SC has them all:
<kbd>![image](https://user-images.githubusercontent.com/24273478/85874776-7230ba80-b7d3-11ea-8d63-fbf29cbd54a9.png)</kbd>

So, I guess providing a complete one might be preferable:
> Other thing to consider is to release our own Noto Serif CJK somewhere and include it like the Lit3rata font.

Also, I'm not seeing any of the other `Noto Sans Khmer|Maynamar...` - May be because we are blacklisting `Noto Sans` ?

poire-z commented:
> We can change them as we like, so probably makes sense to avoid medium completely.

Somehow, I was testing new fonts this morning and I directly went to copying the medium ones, as often the Regular are too thin. So, that might be a user preference.
Ideally, when medium is available, I guess we should ask to download one of the two sets: (regular + bold) or (medium + bold) :)

You might also want to display after a download "Please restart KOReader to see the fonts you just downloaded" :)
Unless we can provide new ones to crengine, but I really don't want to have to look at that :) Although it might just be a matter for registering the new fonts in the same way we do it at startup:
https://github.com/koreader/koreader/blob/46bf88c57307bddf8fa956a95e6d5089b8a94f75/frontend/document/credocument.lua#L96-L114
and to update the `Font>` menu list. But well, that might be tedious.

pazos commented:
> Some crash, that I didn't get on my previous playing, but after I had removed gbooks.lua to get it regenerated without Noto Serif blacklisted, when I hit "Family":

Fixed

> (OK, I have to manually moved them to the font dir.)

Yup, thats not implented :p . I'm going to do it now because I see no reason to wait.

> But shouldn't it download only regular OR medium ? Downloading both will confuse crengine, which will pick the first one, probably medium ?

Font variants that we want are declared in the `variants` table:

```lua
    variants = {
        ["regular"]            = "regular",
        ["italic"]             = "italic",
        ["500"]                = "medium",
        ["500italic"]          = "medium-italic",
        ["700"]                = "bold",
        ["700italic"]          = "bold-italic",
    },
```

We can change them as we like, so probably makes sense to avoid medium completely.

> Also, I'm not seeing any of the other Noto Sans Khmer|Maynamar... - May be because we are blacklisting Noto Sans ?

They're not available. Regular noto (sans/serif) has support for the following scripts: cyrillic+, devanagari, greek+, latin+, vietnamese.

I unblacklisted all of them to test!

poire-z commented on plugins/fontdownloader.koplugin/main.lua:
```diff
@@ -0,0 +1,519 @@
+        text = _("Set download location"),
+        sub_item_table = {
+            {
+                text = _("Set download location"),
```
This goes directly into the More tools menu. I guess the first one should be `Font downloader`.
Also, it's a bit strange to have it that far away from the `Download more fonts`.

I'm also not sure that is really needed. I would download them in a subdirectory of our fonts/ directory, say `fonts/downloaded/` alongside `fonts/noto/` and `fonts/freefonts/` , so it's easier to clean after you have downloaded a bunch of crappy ones :)
(Unless on Android, the `fonts/` and `fonts/noto/` is not user-accessible ?)

pazos replied:
>  I would download them in a subdirectory of our fonts/ directory

getFontDir uses the platform external font directory, where users are expected to put fonts. I thought it would be handy to have a way to override that but probably we can avoid this option.

pazos commented:
> On the first hit on that menu, I see a PING in the logs and nothing else - and the gbooks.lua is created. Only on next menu hit I see the popup. Is that ok ?

Yeah, it needs #6314 to work on the emulator.  Action requires internet so it is scheduled until the network connected.

poire-z commented:
I tested a few of the Featured fonts, and there's something a bit sad: Google Fonts does not really ship the best versions of the fonts...
For example, I downloaded Source Serif Pro: only Regular + Bold as TTF (no small caps) - while at https://github.com/adobe-fonts/source-serif-pro/releases/tag/3.001R, you get some additional Italic, and all in OTF format (with small caps).
Also, Gentium Book Basic were 155 Kb, while the version I already had was > 200 Kb - both TTF. The Windows app that gives info about the font when you click on it gives the exact same metadata for both...
I wouldn't really trust Google fonts for fonts then :) may be only for some kind of preview, but if the preview hides some features, what's the point ? :)

pazos commented:
> I tested a few of the Featured fonts, and there's something a bit sad: Google Fonts does not really ship the best versions of the fonts...

In a ideal world the featured fonts would be all self-hosted and approved by our font-picky masters :)

I played a bit with the idea of repositories. Each one with a base url and an optional api key. Setting a repository would be as simple as inserting the url (and key) and setting a name. Each url should give a json encoded response with a list of fonts (with the same format that gfonts, but that can be changed if we find a better format for them).

Anyways, what google provides is a fast and simple way of downloading fonts. Specially useful for new users.

Frenzie commented:
> adobe-fonts

'nuff said.

NiLuJe commented:
Yeah, the GFonts DB is rarely updated, and will generally prefer hinted or autohinted TTFs.

But, it's there ;p.

FWIW, I also build (some of the) Adobe fonts myself, but I don't actually use 'em in practice, so I don't quite recall which one I do build and what I tweaked (but at the very least, I probably tweaked the vertical metrics).

Galunid commented on plugins/fontdownloader.koplugin/main.lua:
```diff
@@ -0,0 +1,535 @@
+    end
+
+    -- returns true if modification time is equal or less than 7 days old
+    local ok_stamp = function(timestamp)
+        if not timestamp then return false end
+        local current = os.date(self.timestamp_format)
+        if tonumber(current) > timestamp + 7 then return false end
+        return true
+    end
```
```lua
if tonumber(current) > timestamp + 7 then return false end
```
Is this conversion correct? For example:
`current = 20200701` and `timestamp = 20200630`
I assume the comparison would look like `current(20200701) > timestamp(20200637)`. This would return false, when in fact it should be true

pazos replied:
Good catch!

Galunid commented:
Not sure if it's something on my end, but there is a duplicate menu entry(when reader is open, Font -> Font Settings). I cloned the repo and build emulator from scratch. Branch was `gfonts`.
![fonts_download](https://user-images.githubusercontent.com/10298730/85928223-a8da0400-b8ab-11ea-9648-e7edf2fa6b0f.png)

pazos commented:
> Not sure if it's something on my end, but there is a duplicate menu entry(when reader is open, Font -> Font Settings). I cloned the repo and build emulator from scratch. Branch was `gfonts`.

Nope, its not on your end. @poire-z can reproduce in https://github.com/koreader/koreader/pull/6299#issuecomment-650234998. I cannot. I always see one entry, no matter how many books I open. 

Steps to reproduce would be awesome.

Galunid commented:
```
git clone https://github.com/pazos/koreader
cd koreader
git checkout gfonts
./kodev fetch-thirdparty
koreader_run_current
```

`alias koreader_run_current='docker run -e DISPLAY=$DISPLAY --net host --volume="$HOME/.Xauthority:/home/ko/.Xauthority:rw" -v /tmp/.X11-unix:/tmp/.X11-unix -v "$(pwd):/home/ko/koreader" -it koreader/koappimage:latest bash -c "source ~/.bashrc && pushd koreader && ./kodev run"'`
I used latest koreader docker image. Other than that, I started Romeo and Juliet and went to `Font -> Font Settings`.
By the way, amazing plugins, both this and Calibre one :)

poire-z commented:
> Seems like FontDownloader:addToMainMenu was called twice, right after I opened menu(when book was already opened) but I have no idea why.

If I remove this, I see it only once:
https://github.com/koreader/koreader/blob/37cf106fce5964b51cf49054eb2a2f3f03ef3adb/frontend/apps/reader/modules/readermenu.lua#L253-L260

I guess this is always called on the emulator (and @pazos is not using the emulator in debug mode?) and build a uselesss mock_menu_item - but as a side effect, it executes your hack to the change_font submenu twice. So, I guess you'll have to check that you have not already done it to avoid the duplicate.
(I remember I once had to handle that and put an `id=` to my menu, or something like that - can't find what it was now :/)

Btw, another pairs/ipairs messup in readermenu: if we use table.insert(), it's an array and we should use ipairs:
https://github.com/koreader/koreader/blob/37cf106fce5964b51cf49054eb2a2f3f03ef3adb/frontend/apps/reader/modules/readermenu.lua#L417-L418
https://github.com/koreader/koreader/blob/37cf106fce5964b51cf49054eb2a2f3f03ef3adb/frontend/apps/reader/modules/readermenu.lua#L134
https://github.com/koreader/koreader/blob/37cf106fce5964b51cf49054eb2a2f3f03ef3adb/frontend/apps/reader/modules/readermenu.lua#L256

poire-z commented on plugins/fontdownloader.koplugin/main.lua:
```diff
@@ -0,0 +1,527 @@
+    }
+    -- inject in font menu for CRE documents
+    if self.ui.document and self.ui.document.provider == "crengine" then
+        local menu_orig = menu_items.change_font.sub_item_table[1].sub_item_table
+        local position = #menu_orig - 1 -- above fallback fonts
+        local menu = menu_orig
+        table.insert(menu, position, {
+            text = _("Download more fonts"),
+            callback = function()
+                self:frontpage()
+            end,
+        })
+        menu_items.change_font.sub_item_table[1].sub_item_table = menu
```
Here, `menu_items.change_font.sub_item_table[1].sub_item_table`, `menu_orig` and `menu` are all the same object - so no need for this kind of swap :)
```diff
     -- inject in font menu for CRE documents
     if self.ui.document and self.ui.document.provider == "crengine" then
-        local menu_orig = menu_items.change_font.sub_item_table[1].sub_item_table
-        local position = #menu_orig - 1 -- above fallback fonts
-        local menu = menu_orig
+        local menu = menu_items.change_font.sub_item_table[1].sub_item_table
+        local position = #menu - 1 -- above fallback fonts
         table.insert(menu, position, {
             text = _("Download more fonts"),
             callback = function()
                 self:frontpage()
             end,
         })
-        menu_items.change_font.sub_item_table[1].sub_item_table = menu
     end
```

poire-z commented:
> (I remember I once had to handle that and put an id= to my menu, or something like that - can't find what it was now :/)

OK, not exactly that, but I added a flag so that I can remove it before inserting it again:
https://github.com/koreader/koreader/blob/37cf106fce5964b51cf49054eb2a2f3f03ef3adb/frontend/apps/reader/modules/readermenu.lua#L170-L183

pazos commented:
Thanks!!!! That's it! I removed dbg:turnOn() on reader.lua. Now I can reproduce :)

Galunid commented:
Can confirm, only one entry :)

Frenzie commented:
@Galunid Sidenote, and there's no wrong method, but these days I generally do what I described in https://github.com/koreader/koreader/pull/6289 to test PRs, only bothering with branches if I want to actively change/push something slightly more complicated.

pazos commented:
I found Fontsquirrel api, which adds other 1000 fonts to the table. It is not going to happen for this PR but maybe someday. So I split code in repos. That makes easier to add new repos on the future.

NiLuJe commented:
Random comment to mention that I updated Literata to Literata 3.002 (which shouldn't change much, but, hey, bigger numbers are better, right? :D).

pazos commented:
> Random comment to mention that I updated Literata to Literata 3.002 (which shouldn't change much, but, hey, bigger numbers are better, right? :D).

If you want to feed your static/dynamic json feel free to give me an url and an optional api key ;). Actually your "repo" is local, so needs to be updated with each new version: https://github.com/koreader/koreader/pull/6299/files#diff-22fafaaabcee719340bb047cdf2df6ea.

NiLuJe commented:
@pazos: I didn't update the URL, though, I just started *also* duplicating them in versioned subfolders, but the top-level ones point to the updated version ;).

i.e.,

```
┌─(niluje@illyria:pts/10)────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────(/var/tmp/niluje/staging/cloud/3.002)─┐
└─(0.30:33%:04:54:%)── swift list koreader-pub                                                                                                                                                                                                                                                                                                         ──(Fri, Jul 31)─┘
3.0/Lit3rata-Bold.otf
3.0/Lit3rata-BoldItalic.otf
3.0/Lit3rata-Italic.otf
3.0/Lit3rata-Regular.otf
3.002/Lit3rata-Bold.otf
3.002/Lit3rata-BoldItalic.otf
3.002/Lit3rata-Italic.otf
3.002/Lit3rata-Regular.otf
Lit3rata-Bold.otf
Lit3rata-BoldItalic.otf
Lit3rata-Italic.otf
Lit3rata-Regular.otf
```

Frenzie commented:
@pazos What's the status on this one, postpone to 2020.10?

pazos commented:
> What's the status on this one, postpone to 2020.10?

@Frenzie: it is almost ready to go 

Just need to address: https://github.com/koreader/koreader/pull/6299#pullrequestreview-451939475 and https://github.com/koreader/koreader/pull/6299#discussion_r446271774.

And test on android, which is the only platform with more than one folder for user fonts.

Probably some things can be improved later (like https://github.com/koreader/koreader/pull/6299#issuecomment-650251096), but since it is a new plugin I wouldn't mind to include it in 2020.09, even if it is a little bit more buggier than I would want :smile:
